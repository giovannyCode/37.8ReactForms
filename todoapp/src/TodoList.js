import React, { useState } from "react";
import Todo from "./Todo.js";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm.js";

function TodoList() {
  const arraytodos = [
    { id: 1, task: "Learn React basics" },
    { id: 2, task: "Go grocery shopping" },
    { id: 3, task: "Work out at the gym" },
    { id: 4, task: "Read a new book" },
    { id: 5, task: "Prepare a report for work" },
    { id: 6, task: "Plan a weekend trip" },
    { id: 7, task: "Cook a new recipe" },
  ];
  const [todos, setTodos] = useState(arraytodos);

  /** Add new item object to array. */
  const addTodo = (todo) => {
    let newTodo = { ...todo, id: uuid() };
    setTodos((todos) => [...todos, newTodo]);
  };
  const deleteTodo = (id) => {
    const updatedTodos = todos.filter((todo) => todo.id !== id);
    setTodos(updatedTodos);
  };
  return (
    <div>
      {<NewTodoForm addTodo={addTodo} />}
      <br></br>
      {todos.map((todo) => (
        <Todo
          key={todo.id}
          task={todo.task}
          onRemove={deleteTodo}
          id={todo.id}
        />
      ))}
    </div>
  );
}

export default TodoList;
