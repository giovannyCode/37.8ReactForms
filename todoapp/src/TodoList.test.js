import React from "react";
import {
  render,
  fireEvent,
  screen,
  getByText,
  within,
  queryByText,
} from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("Add a new Box", function () {
  const { container } = render(<TodoList />);
  const taskInput = container.querySelector("#task");
  fireEvent.change(taskInput, { target: { value: "Steam Vegetables" } });
  const addButton = getByText(container, "Add a new Todo!");
  fireEvent.click(addButton);
  const newTask = screen.getByText("Steam Vegetables");
  expect(newTask).toBeInTheDocument();
});

it("Deletes Todo", function () {
  const { container } = render(<TodoList />);

  const todoDiv = screen.getByText("Plan a weekend trip").closest(".todo");
  const deleteButton = within(todoDiv).getByRole("button", { name: /x/i });

  fireEvent.click(deleteButton);
  const taskPlanaWeekendTrip = queryByText(container, "Plan a weekend trip");

  expect(taskPlanaWeekendTrip).not.toBeInTheDocument();
});
