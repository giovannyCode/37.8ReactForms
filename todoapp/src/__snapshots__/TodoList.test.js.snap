// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        task:
      </label>
      <input
        id="task"
        name="task"
        value=""
      />
      <button>
        Add a new Todo!
      </button>
    </form>
    <br />
    <div
      class="todo"
    >
      <span>
        Learn React basics
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Go grocery shopping
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Work out at the gym
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Read a new book
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Prepare a report for work
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Plan a weekend trip
      </span>
      <button>
        X
      </button>
    </div>
    <div
      class="todo"
    >
      <span>
        Cook a new recipe
      </span>
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
